larsim/EventGenerator/MultiPart/MultiPartRain_module.cc

```diff
@@ -0,0 +1,547 @@
+////////////////////////////////////////////////////////////////////////
+// Class:       MultiPartRain
+// Module Type: producer
+// File:        MultiPartRain_module.cc
+//
+// Generated at Tue Dec 13 15:48:59 2016 by Kazuhiro Terao using artmod
+// from cetpkgsupport v1_11_00.
+////////////////////////////////////////////////////////////////////////
+
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+//#include "art/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+#include "cetlib/pow.h"
+#include <cmath>
+#include <memory>
+#include <vector>
+
+#include "CLHEP/Random/RandFlat.h"
+#include "CLHEP/Random/RandGauss.h"
+#include "CLHEP/Random/RandGeneral.h"
+
+#include "TRandom.h"
+#include "larcore/CoreUtils/ServiceUtil.h"
+#include "larcore/Geometry/Geometry.h"
+#include "larcoreobj/SimpleTypesAndConstants/PhysicalConstants.h"
+#include "larcoreobj/SummaryData/RunData.h"
+#include "nurandom/RandomUtils/NuRandomService.h"
+
+#include "nusimdata/SimulationBase/MCParticle.h"
+#include "nusimdata/SimulationBase/MCTruth.h"
+
+#include "TDatabasePDG.h"
+#include "TLorentzVector.h"
+
+struct PartGenParam {
+  std::vector<int> pdg;
+  std::vector<double> mass;
+  std::array<size_t, 2> multi;
+  std::array<double, 2> kerange;
+  bool use_mom;
+  bool direct_inward;
+  double weight;
+};
+
+class MultiPartRain;
+
+class MultiPartRain : public art::EDProducer {
+public:
+  explicit MultiPartRain(fhicl::ParameterSet const& p);
+  // The destructor generated by the compiler is fine for classes
+  // without bare pointers or other resource use.
+  ~MultiPartRain();
+
+  // Plugins should not be copied or assigned.
+  MultiPartRain(MultiPartRain const&) = delete;
+  MultiPartRain(MultiPartRain&&) = delete;
+  MultiPartRain& operator=(MultiPartRain const&) = delete;
+  MultiPartRain& operator=(MultiPartRain&&) = delete;
+
+  // Required functions.
+  void produce(art::Event& e) override;
+
+  void beginRun(art::Run& run) override;
+
+  void GenPosition(double& x, double& y, double& z);
+
+  std::array<double, 3U> extractDirection() const;
+  void GenMomentum(const PartGenParam& param,
```

Comment:
It would be better for `GenPosition` to return a `TVector3` corresponding to the momentum.